File: frontend/src/core/activity/reducer.ts
/**
 * Fold activity transitions into the rows the rail shows.
 *
 * The same three rules the backend reducer enforces, for the same reasons:
 * replay is not new information, a settled row is closed for good, and a row
 * must be opened before it can be updated. They are restated here rather than
 * trusted from the server because the browser sees a *different* stream — one
 * that reconnects, replays, and joins mid-flight.
 *
 * **Coalescing belongs here, not in the view.** An update that changes neither
 * actor, state, operation, nor lineage returns the *same state object*, so
 * React bails out of the render entirely. Without that, a token-rate stream
 * re-renders the block many times a second to paint an identical spinner — and
 * that is not only wasted work: a remounted `animate-spin` restarts its
 * rotation, so the one moving thing on screen stutters.
 */

import {
  type ActivityEvent,
  type ActivityProjection,
  type ActivityRow,
  type ActivityTimelineEntry,
  EMPTY_ACTIVITY,
  isTerminalActivityState,
} from "./types";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function optionalString(value: unknown): string | null {
  return typeof value === "string" && value ? value : null;
}

/**
 * Narrow an unknown stream frame to an activity event.
 *
 * Deliberately permissive about *values* and strict about *shape*: the server
 * already validated the vocabulary against a closed field set, so re-checking
 * every enum member here would mean a backend that learns a new state silently
 * loses its rows in every older client. What must be present is the identity a
 * row cannot be folded without.
 */
export function asActivityEvent(frame: unknown): ActivityEvent | null {
  if (!isRecord(frame) || frame.type !== "agent_activity") return null;
  const { activity_id, run_id, transition, state, display_name } = frame;
  if (typeof activity_id !== "string" || !activity_id) return null;
  if (typeof run_id !== "string" || !run_id) return null;
  if (typeof transition !== "string" || !transition) return null;
  if (typeof state !== "string" || !state) return null;
  if (typeof display_name !== "string" || !display_name) return null;
  return frame as unknown as ActivityEvent;
}

interface NormalizedActivityFrame {
  event: ActivityEvent;
  serverSeq: number | null;
  receivedAt: number;
}

function normalizeActivityFrame(
  frame: unknown,
  now: number,
): NormalizedActivityFrame | null {
  if (
    isRecord(frame) &&
    typeof frame.seq === "number" &&
    Number.isInteger(frame.seq) &&
    frame.seq > 0
  ) {
    const event = asActivityEvent(frame.content);
    if (!event) return null;
    const parsedCreatedAt =
      typeof frame.created_at === "string"
        ? Date.parse(frame.created_at)
        : Number.NaN;
    return {
      event,
      serverSeq: frame.seq,
      receivedAt: Number.isNaN(parsedCreatedAt) ? now : parsedCreatedAt,
    };
  }
  const event = asActivityEvent(frame);
  return event ? { event, serverSeq: null, receivedAt: now } : null;
}

function rowFromEvent(
  event: ActivityEvent,
  seq: number,
  serverSeq: number | null,
): ActivityRow {
  const scope = isRecord(event.scope) ? event.scope : {};
  return {
    activityId: event.activity_id,
    runId: event.run_id,
    actorKind: event.actor_kind,
    actorId: event.actor_id,
    displayName: event.display_name,
    state: event.state,
    operation: event.operation,
    parentActivityId: optionalString(event.parent_activity_id),
    dispatcherActivityId: optionalString(event.dispatcher_activity_id),
    cycleId: optionalString((scope as Record<string, unknown>).cycle_id),
    stage: optionalString((scope as Record<string, unknown>).stage),
    taskId: optionalString((scope as Record<string, unknown>).task_id),
    seq,
    serverSeq,
  };
}

function timelineEntry(
  event: ActivityEvent,
  seq: number,
  serverSeq: number | null,
  receivedAt: number,
): ActivityTimelineEntry {
  const scope = isRecord(event.scope)
    ? (event.scope as Record<string, unknown>)
    : {};
  return {
    seq,
    serverSeq,
    runId: event.run_id,
    activityId: event.activity_id,
    displayName: event.display_name,
    actorKind: event.actor_kind,
    transition: event.transition,
    state: event.state,
    operation: event.operation,
    cycleId: optionalString(scope.cycle_id),
    stage: optionalString(scope.stage),
    receivedAt,
  };
}

/** Fold one frame. Returns the *same* projection when nothing changed. */
export function applyActivityEvent(
  projection: ActivityProjection,
  frame: unknown,
  now: number = Date.now(),
): ActivityProjection {
  const normalized = normalizeActivityFrame(frame, now);
  if (!normalized) return projection;
  const { event, serverSeq, receivedAt } = normalized;

  const existing = projection.rows[event.activity_id];

  if (
    existing?.serverSeq !== null &&
    existing?.serverSeq !== undefined &&
    serverSeq !== null &&
    serverSeq <= existing.serverSeq
  ) {
    return projection;
  }

  // A closed row is a fact about work that ended. Refusing here is what stops a
  // replayed opening from putting a finished actor back on screen.
  if (existing && isTerminalActivityState(existing.state)) return projection;

  if (event.transition === "started") {
    // A replayed opening for a row already open changes nothing; taking the
    // newer payload would let a stale replay overwrite a live state.
    if (existing) return projection;
  } else if (!existing) {
    // An update or close for a row this client never saw open. Backfill
    // supplies the opening; a synthesised row would misreport its lineage.
    return projection;
  }

  const seq = projection.seq + 1;
  const next = rowFromEvent(
    event,
    seq,
    serverSeq ?? existing?.serverSeq ?? null,
  );

  if (
    existing?.state === next.state &&
    existing.operation === next.operation &&
    existing.displayName === next.displayName &&
    existing.parentActivityId === next.parentActivityId &&
    existing.dispatcherActivityId === next.dispatcherActivityId
  ) {
    // A newer durable sequence still advances the replay watermark, but does
    // not become a visible history step or change active-leaf recency.
    return serverSeq !== null && serverSeq !== existing.serverSeq
      ? {
          ...projection,
          rows: {
            ...projection.rows,
            [event.activity_id]: {
              ...existing,
              serverSeq,
            },
          },
        }
      : projection;
  }

  const timeline = [
    ...projection.timeline,
    timelineEntry(event, seq, serverSeq, receivedAt),
  ];
  return {
    rows: { ...projection.rows, [event.activity_id]: next },
    order: existing
      ? projection.order
      : [...projection.order, event.activity_id],
    timeline,
    seq,
  };
}

export function reduceActivityEvents(
  frames: readonly unknown[],
  projection: ActivityProjection = EMPTY_ACTIVITY,
  now: number = Date.now(),
): ActivityProjection {
  return frames.reduce<ActivityProjection>(
    (current, frame) => applyActivityEvent(current, frame, now),
    projection,
  );
}

/** Rows still working, in the order they opened. */
export function activeRows(
  projection: ActivityProjection,
): readonly ActivityRow[] {
  const rows: ActivityRow[] = [];
  for (const id of projection.order) {
    const row = projection.rows[id];
    if (row && !isTerminalActivityState(row.state)) rows.push(row);
  }
  return rows;
}

/**
 * The active rows with no active child.
 *
 * "Who is working now" wants the innermost answer: a supervisor that has
 * dispatched a stage is waiting, and naming it describes the tree rather than
 * the work.
 */
export function activeLeaves(
  projection: ActivityProjection,
): readonly ActivityRow[] {
  const active = activeRows(projection);
  const parents = new Set(
    active
      .map((row) => row.parentActivityId)
      .filter((id): id is string => Boolean(id)),
  );
  return active.filter((row) => !parents.has(row.activityId));
}

/** The chain from a row up to its root, nearest ancestor first. */
export function dispatcherChain(
  projection: ActivityProjection,
  row: ActivityRow | undefined,
): readonly ActivityRow[] {
  const chain: ActivityRow[] = [];
  const seen = new Set<string>();
  let current = row?.parentActivityId
    ? projection.rows[row.parentActivityId]
    : undefined;
  while (current && !seen.has(current.activityId)) {
    seen.add(current.activityId);
    chain.push(current);
    current = current.parentActivityId
      ? projection.rows[current.parentActivityId]
      : undefined;
  }
  return chain;
}

/**
 * Settle every still-open row.
 *
 * A run ends whether or not each actor said so — a cancelled run, a lost lease,
 * a dropped connection. `interrupted` is the honest terminal: the work neither
 * finished nor demonstrably failed, it stopped being observed.
 */
export function closeOpenRows(
  projection: ActivityProjection,
  state: ActivityRow["state"] = "interrupted",
): ActivityProjection {
  const open = activeRows(projection);
  if (open.length === 0) return projection;
  const rows = { ...projection.rows };
  for (const row of open) {
    rows[row.activityId] = { ...row, state };
  }
  return { ...projection, rows };
}

File: backend/packages/harness/deerflow/runtime/activity/reducer.py
"""Fold a stream of activity transitions into the set of rows a reader sees.

This is the consumer half of the contract the producer already obeys, written
once here so the durable projection, a reconnecting client, and the eventual
rail cannot each invent their own answer to "what does this sequence mean?".
It is pure: no store, no clock, no IO.

Three rules carry the weight, and each exists because of a specific way a live
stream lies to whoever is reading it.

**Replay is not new information.** A reconnect re-delivers events the reader has
already folded, and a durable backfill overlaps the live tail by design. So
reduction is idempotent: folding a sequence twice, or folding a prefix and then
the whole thing, yields the same rows. Nothing here counts, appends, or
accumulates.

**A settled row is closed for good.** Once a terminal transition lands, later
frames for that id are refused — including another ``started``. A row that could
reopen is a spinner that returns after the work finished, and the id is the
thing being reused; the producer already mints a fresh id per invocation, so a
repeated id is either a replay (harmless, refused) or a bug (visible, refused).

**A row must be opened before it can be updated.** An ``updated`` frame for an
unknown id is dropped rather than promoted into a row. Mid-stream joins are
normal — a client subscribing late sees updates whose opening it missed — and
inventing a row from one would show work whose beginning, parent, and actor the
reader never learned. Backfill supplies the opening; guessing does not.
"""

from collections.abc import Iterable, Mapping
from dataclasses import dataclass, replace
from typing import Any, Final

from deerflow.runtime.activity.envelope import parse_activity_event
from deerflow.runtime.activity.vocabulary import (
    TERMINAL_TRANSITIONS,
    ActivityState,
    ActivityTransition,
    ActorKind,
)

_TERMINAL_TRANSITION_VALUES: Final[frozenset[str]] = frozenset(member.value for member in TERMINAL_TRANSITIONS)


@dataclass(frozen=True, slots=True)
class ActivityRow:
    """One actor's current presence, as folded from its transitions."""

    activity_id: str
    run_id: str
    actor_kind: ActorKind
    actor_id: str
    display_name: str
    state: ActivityState
    operation: str
    parent_activity_id: str | None = None
    dispatcher_activity_id: str | None = None
    cycle_id: str | None = None
    stage: str | None = None
    task_id: str | None = None
    last_seq: int | None = None

    @property
    def settled(self) -> bool:
        """Whether this row has closed and can no longer change."""
        return self.state in {ActivityState.COMPLETED, ActivityState.FAILED, ActivityState.CANCELLED, ActivityState.INTERRUPTED}

    @property
    def active(self) -> bool:
        return not self.settled


def _persisted_payload(event: object) -> tuple[dict[str, Any] | None, int | None]:
    """Return a canonical payload and its durable sequence, when present."""
    if isinstance(event, Mapping):
        seq = event.get("seq")
        content = event.get("content")
        if isinstance(seq, int) and not isinstance(seq, bool) and seq > 0 and isinstance(content, Mapping):
            return parse_activity_event(content), seq
    return parse_activity_event(event), None


def _row_from_payload(payload: Mapping[str, Any], *, seq: int | None = None) -> ActivityRow:
    scope = payload["scope"]
    return ActivityRow(
        activity_id=payload["activity_id"],
        run_id=payload["run_id"],
        actor_kind=ActorKind(payload["actor_kind"]),
        actor_id=payload["actor_id"],
        display_name=payload["display_name"],
        state=ActivityState(payload["state"]),
        operation=payload["operation"],
        parent_activity_id=payload["parent_activity_id"],
        dispatcher_activity_id=payload["dispatcher_activity_id"],
        cycle_id=scope.get("cycle_id"),
        stage=scope.get("stage"),
        task_id=scope.get("task_id"),
        last_seq=seq,
    )


def apply_activity_event(rows: Mapping[str, ActivityRow], event: object) -> dict[str, ActivityRow]:
    """Fold one event into ``rows`` and return the new mapping.

    Immutable by construction — the input mapping is never modified, so a caller
    holding an earlier snapshot keeps it. Insertion order is the order rows were
    first opened, which is the order a reader watched the work begin in.
    """
    payload, seq = _persisted_payload(event)
    if payload is None:
        return dict(rows)

    activity_id = payload["activity_id"]
    existing = rows.get(activity_id)
    transition = payload["transition"]

    if existing is not None and seq is not None and existing.last_seq is not None and seq <= existing.last_seq:
        # A reconnect or overlapping page can deliver an older durable record
        # after a newer one. Sequence is server authority; arrival order is not.
        return dict(rows)

    if existing is not None and existing.settled:
        # Refused, not applied: a closed row is a fact about work that ended.
        return dict(rows)

    if transition == ActivityTransition.STARTED.value:
        if existing is not None:
            # A replayed opening for a row already open changes nothing; taking
            # the newer payload would let a stale replay overwrite a live state.
            return dict(rows)
        updated = dict(rows)
        updated[activity_id] = _row_from_payload(payload, seq=seq)
        return updated

    if existing is None:
        # An update or a close for a row this reader never saw open. Backfill
        # supplies the opening; a synthesised row would misreport its lineage.
        return dict(rows)

    updated = dict(rows)
    updated[activity_id] = replace(
        existing,
        state=ActivityState(payload["state"]),
        operation=payload["operation"],
        last_seq=seq if seq is not None else existing.last_seq,
    )
    return updated


def reduce_activity_events(events: Iterable[object], *, rows: Mapping[str, ActivityRow] | None = None) -> dict[str, ActivityRow]:
    """Fold a sequence of events, optionally continuing from an earlier snapshot."""
    current: dict[str, ActivityRow] = dict(rows or {})
    for event in events:
        current = apply_activity_event(current, event)
    return current


def active_rows(rows: Mapping[str, ActivityRow]) -> list[ActivityRow]:
    """Return the rows still working, in the order they opened."""
    return [row for row in rows.values() if row.active]


def active_leaves(rows: Mapping[str, ActivityRow]) -> list[ActivityRow]:
    """Return the active rows that have no active child.

    What a reader wants from "who is working now" is the innermost answer: a
    supervisor that has dispatched a stage is waiting, and naming it instead of
    the stage describes the tree rather than the work.
    """
    parents = {row.parent_activity_id for row in rows.values() if row.active and row.parent_activity_id}
    return [row for row in active_rows(rows) if row.activity_id not in parents]


def close_open_rows(rows: Mapping[str, ActivityRow], *, state: ActivityState = ActivityState.INTERRUPTED) -> dict[str, ActivityRow]:
    """Settle every still-open row.

    A run ends whether or not each actor got to say so — a lease loss, a hard
    cancellation, or a process death leaves rows that nothing will ever close.
    The honest terminal there is ``interrupted``: the work neither finished nor
    demonstrably failed, it stopped being observed.
    """
    return {activity_id: (row if row.settled else replace(row, state=state)) for activity_id, row in rows.items()}
